### PR TITLE
Bugfix/Fix songHash exposure (#39)

### LINKED ISSUE #32: [Feature Request] Get beat saver song code?
Ive read through the source and I have no clue how it works so IDK if you could find a way to have the plugin get the current song code

### LINKED ISSUE #39: Pass through songHash in StatusManager.cs
## Feature request
I need the songHash that is already present (Plugin.cs) being passed to the response beatmap json in StatusManager.cs.

## Motivation
I want to get the !bsr ID in my adaptation of the SongInfo Overlay by requesting a json response from https://beatsaver.com/api/maps/by-hash/$data.songHash .

A mockup is already working, all i need is the songHash by you guys :)

### PR DESCRIPTION
**Pull request by ErisApps**
This PR should resolve #32 (kinda) and also resolves #39 

It exposes the songHash and fixes the extraction of the songHash as the format has changed.

## Benchmarks of extraction code
The way of extraction has also been benchmarked and optimized for the most common use case (being custom songs) using the following piece of code
```csharp
using System;
using System.Text.RegularExpressions;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Environments;
using BenchmarkDotNet.Exporters;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;

namespace Benchmarks
{
	class Program
	{
		static void Main(string[] args)
		{
			BenchmarkRunner.Run<SongHashBenchmark>(DefaultConfig.Instance
				.AddJob(Job.Default.WithRuntime(ClrRuntime.Net461)
					.WithJit(Jit.Default))
				.AddDiagnoser(MemoryDiagnoser.Default)
				.AddExporter(BenchmarkReportExporter.Default, HtmlExporter.Default, MarkdownExporter.GitHub)
				.KeepBenchmarkFiles());
		}
	}
	
	public class SongHashBenchmark
	{
		[Params(
			"AngelVoices",
			"custom_level_DD12E126F4AB384AF7D20154416DCCEEB173380E",
			"custom_level_DD12E126F4AB384AF7D20154416DCCEEB173380E WIP")]
		public string LevelId;
		
		public readonly Regex CompiledSongHashRegex = new Regex("^custom_level_(.*)(?<! WIP)$", RegexOptions.Compiled);

		[Benchmark(Baseline = true)]
		public string ExtractUsingSubstring()
		{
			if (LevelId.StartsWith("custom_level_") && !LevelId.EndsWith(" WIP"))
			{
				return LevelId.Substring(13, LevelId.Length - 13);
			}
			else
			{
				return null;
			}
		}

		[Benchmark]
		public string ExtractUsingOnlyReplace()
		{
			if (LevelId.StartsWith("custom_level_") && !LevelId.EndsWith(" WIP"))
			{
				return LevelId.Replace("custom_level_", string.Empty);
			}
			else
			{
				return null;
			}
		}

		[Benchmark]
		public string ExtractUsingOnlySplit()
		{
			if (LevelId.StartsWith("custom_level_") && !LevelId.EndsWith(" WIP"))
			{
				return LevelId.Split(new[] {"custom_level_"}, StringSplitOptions.RemoveEmptyEntries)[0];
			}
			else
			{
				return null;
			}
		}

		[Benchmark]
		public string ExtractUsingRegex()
		{
			var match = CompiledSongHashRegex.Match(LevelId);
			return match.Success ? match.Groups[1].Value : null;
		}
	}
}
```

## With the following results

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.815 (1909/November2018Update/19H2)
AMD Ryzen 9 3900X, 1 CPU, 24 logical and 12 physical cores
.NET Core SDK=3.1.201
  [Host]     : .NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT
  Job-NRJDXI : .NET Framework 4.8 (4.8.4150.0), X64 RyuJIT

Jit=Default  Runtime=.NET 4.6.1  

|                  Method |              LevelId |        Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------ |--------------------- |------------:|----------:|----------:|------:|--------:|-------:|------:|------:|----------:|
|   ExtractUsingSubstring |          AngelVoices |    76.83 ns |  1.549 ns |  1.449 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| ExtractUsingOnlyReplace |          AngelVoices |    76.97 ns |  1.167 ns |  1.092 ns |  1.00 |    0.03 |      - |     - |     - |         - |
|   ExtractUsingOnlySplit |          AngelVoices |    76.76 ns |  1.560 ns |  1.670 ns |  1.00 |    0.03 |      - |     - |     - |         - |
|       ExtractUsingRegex |          AngelVoices |    64.36 ns |  0.933 ns |  0.873 ns |  0.84 |    0.02 |      - |     - |     - |         - |
|                         |                      |             |           |           |       |         |        |       |       |           |
|   ExtractUsingSubstring | custo(...)3380E [53] |   191.98 ns |  3.082 ns |  2.883 ns |  1.00 |    0.00 | 0.0675 |     - |     - |     112 B |
| ExtractUsingOnlyReplace | custo(...)3380E [53] |   315.82 ns |  4.415 ns |  4.130 ns |  1.65 |    0.03 | 0.0672 |     - |     - |     112 B |
|   ExtractUsingOnlySplit | custo(...)3380E [53] |   348.05 ns |  2.281 ns |  1.905 ns |  1.82 |    0.03 | 0.4201 |     - |     - |     698 B |
|       ExtractUsingRegex | custo(...)3380E [53] |   294.72 ns |  5.490 ns |  5.135 ns |  1.54 |    0.04 | 0.3185 |     - |     - |     530 B |
|                         |                      |             |           |           |       |         |        |       |       |           |
|   ExtractUsingSubstring | custo(...)E WIP [57] |   178.54 ns |  3.234 ns |  3.026 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| ExtractUsingOnlyReplace | custo(...)E WIP [57] |   177.78 ns |  2.878 ns |  2.692 ns |  1.00 |    0.01 |      - |     - |     - |         - |
|   ExtractUsingOnlySplit | custo(...)E WIP [57] |   178.85 ns |  2.917 ns |  2.728 ns |  1.00 |    0.01 |      - |     - |     - |         - |
|       ExtractUsingRegex | custo(...)E WIP [57] | 1,366.01 ns | 18.618 ns | 17.415 ns |  7.65 |    0.11 |      - |     - |     - |         - |

### CODE REVIEW

**Comment by opl-** [comment]
- Are these all the possible values for `LevelId`?
- Is the `custom_level_` prefix generated by a mod while all the vanilla songs just use "normal" names like `AngelVoices`?
- Should we ignore the WIP part (where does it come from)?
- Should we expose both the hash and the full `levelId`?

**Comment by ErisApps** [comment]
1. & 2. As far as I've tested, those are basically all the case I've come across (but feel free to test yourself 🙂 )
- For OST songs like AngelVoices and such. The LevelId won't include a songHash, like the custom levels do, but it simply be the SongName without spaces as far as I know.
- The `custom_level_` doesn't seem to be added by another mod as custom levels are supposedly part of the basegame. SongDataCore doesn't influence this either.
3. I'm not certain what all the cases are when the ` WIP` suffix is included, but it is at least included when the a song is put in the `Beat Saber_Data/CustomWIPLevels` folder.
Whether we should ignore it or not is up to you. I'll leave that up to your decision, although it's probably also dependent on the next answer.
4. While I don't really think that exposing the full levelId has any "real-world" benefit, it might be useful to include the full levelId as we're still leaving out information that might be "essential" to others. Other than that, as we're just copying pointers as a string is just a reference so I don't think it has any meaningfull performance impact aside from the serialization when sending the message over the WebSocket.
TL;DR: I don't mind including it.

**Comment by opl-** [comment]
Hm. I'd say let's expose `levelId` on top of `songHash`, just in case someone finds a use for it, and leave `songHash` empty if the song doesn't have a hash. This should allow people to easily determine if it makes sense to send a request to BeatSaver by just checking if `songHash` is `null`.

I'd appreciate it if you could please update the `protocol.md` accordingly.

**Comment by ErisApps** [comment]
I've updated the benchmarks and results of the logic behind the songHash as that has been changed.
Protocol.md is updated as well.

**Comment by opl-** [comment]
:ok_hand: Thank you.
